text/2022-01-12-rpc-through-unix-socket.md

```diff
@@ -0,0 +1,242 @@
+
+This RFC **don't** touch anything about application layer rpc protocol.
+
+## Motivation
+
+The chaos-daemon communicates with `rs-tproxy`, `toda`... with stdin/stdout.
+However, the `stdin` and `stdout` is not good enough for communicating. We
+should take care of the log print (make sure that every log falls into
+`stderr`), blocking process (through the `bpm/buffer`, which is hard to
+understand), and establish a mordern rpc application layer on such a fragile
+bottom layer.
+
+This problem can be solved by opening an extra socket or pipe to communicate
+with a subprocess. As it's hard to move a named socket to another mount
+namespace, the abstract unix socket is the best choice.
+
+## Detailed Design
+
+When it comes to the communicating between chaos-daemon and its subprocess, we
+have three components need to describe:
+
+1. Chaos Daemon
+2. nsexec
+3. The subprocess
+
+### Chaos Daemon
+
+The bpm should enable the Chaos Daemon to pass an extra file (which is the unix
+socket fd) to the subprocess.
+
+```go
+func (b *ProcessBuilder) WithSocket() *ProcessBuilder {
+    b.WithSocket = true
+    return b
+}
+```
+
+While building this process, this field should be set to the `ExtraFiles`:
+
+```go
+// Build builds the process
+func (b *ProcessBuilder) Build() *ManagedProcess {
+    ...
+    if b.WithSocket {
+        rawListener, err := net.Listen("unix", fmt.Sprintf("@chaos-daemon-%s", *b.identifier))
+        listener := rawListener.(*net.UnixListener)
+        listenSocket, err := listener.File()
+        command.ExtraFiles = append(command.ExtraFiles, listenSocket)
+    }
+    ...
+}
+```
+
+Then the chaos daemon can send command by dialing this abstract socket:
+
+```go
+httpc := http.Client{
+    Transport: &http.Transport{
+        DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+            return net.Dial("unix", fmt.Sprintf("@chaos-daemon-%s", *b.identifier))
+        },
+    },
+}
+```
+
+The name of the abstract unix socket is `@chaos-daemon-{b.identifier}`, which
+means communicating is only avaiable for process with an identifier. I think
+it's acceptable. As as alternative, we can generate an UUID as the name for
+every process, but I prefer the identifier (for the convenience of debug)
+
+ **Make sure that every LISTENING sockets are closed after the command has
+started, or the parent will be blocked by dialing dead children**.
+
+If every listening fd is closed, further request will get an error: `dial unix
+@xxxx: connect: connection refused`.
+
+### nsexec
+
+The newest version of `nsexec` already supports passing files to its subprocess,
+with the help of [command-fds](https://github.com/google/command-fds)
+
+### Subprocess
+
+The subprocess, for example `toda`, will need to establish its own transport
+from a raw fd. In go, the raw fd can be converted to a `os.File` directly:
+
+```go
+s := os.NewFile(3, "socket")
+
+listener, err := net.FileListener(s)
+```
```

Comment:
Open a file with `fd 3` is not OK to me. It would let subprocess have to make changes for adapting this channel, and it's not easy to debug and develop, because it always acquires `fd 3` as an available socket.
 
I prefer to build a socket to socket channel/proxy with alternative 2 with a named socket, and subprocess could know nothing about this channel:

- subprocess could listen on an endpoint `0.0.0.0:54321` with IP `10.0.3.87` in some networkspace, and this tool makes a tunnel in the main processes' somewhere(working directory or `/var/run` or `/tmp`). Then other processes could access this socket file as same as communicating with `10.0.3.87:54321`
- subprocess could listen on a Unix Domain Socket file `/var/run/toda.sock` in some mount namespace , and this tool make a tunnel..(same as above)

Reply:
I think acquiring a special fd is a common solution, as a process always acquires `fd 0`,  `fd 1`, and `fd 2` as stdin, stdout, and stderr. Moreover, building a channel brings much more maintenance costs than the current solution (communicate on stdio).

Reply:
> I wish this tool could be a more common-useful library/binary to tunnel/proxy sockets across network namespace or mount namespace.
> 
> Not coupled with `chaoe-daemon` and `bpm`.
> 
> And only assume that "subprocess" (or target process) is a normal network application which listen on a network address or UNIX domain socket ❤️

We can just implement that tunnel in the `nsexec`.

Reply:
@STRRL  I don't think it would be a big problem to require an extra fd to be unix socket. It could be an argument: e.g. `toda --listen fd://3`, then it will listen on fd 3, or `toda --listen /mnt/some.socket`, it will listen on a named socket for debug.

I really don't expect to manage another process to establish the socket. It would be a big burden for `bpm`.

Reply:
@Hexilee

> We can just implement that tunnel in the `nsexec`.

It's possible to listen on abstract unix socket and proxy them across the network namespace (as different thread can locate in different network namespaces). But doing so on a named socket (like '/mnt/some.sock') would be much harder. Because different thread cannot locate in different mnt namespace (`pthread_create` creates thread with `CLONE_FS`). `clone` manually or spawning a new process are all tiring.

Reply:
If I understand correctly, `fd 3 `could only be "inherited" from its parent process, which is not so flexible as using the named unix socket domain.

> I really don't expect to manage another process to establish the socket. It would be a big burden for bpm.

I also would not like to append more logic into `bpm`. 

How about making a new component that is out of `bpm`?

For example a binary `tunnel`, it would only do "port forwarding" thing from target unix domain socket or network address, it could resolve namespace's thing by itself, and it could be used like 

- `tunnel -p <pid> --upstream unix:///var/run/toda.sock --listen 0.0.0.0:54321`
- `tunnel -p <pid> --upstream <some-ip>:<port> --listen 0.0.0.0:54321`
- `tunnel -p <pid> --upstream  unix:///var/run/toda.sock --listen unix:///var/run/toda-with-identifier.sock`
- other combination...

Or we could go methods like 

```
type Upstream Interface {}
type Sink Interface {}

Tunnel(ctx context.Context, upstream Upstream, sink Sink){
}
```

It does not care about the subprocess's lifecycle, and the subprocess need not be its child process.

> But doing so on a named socket (like '/mnt/some.sock') would be much harder. Because different threads cannot locate in different mnt namespace(pthread_create creates thread with CLONE_FS). clone manually or spawning a new process are all tiring.

We could use `fd://3` inside our `tunnel`, instead of using fd to communicate with the other process directly:

- we would provide two binary: `tunnel` and `sink`
- `tunnel` is always the parent process for `sink`, and using `fd 3` to communicate with `sink`
- `sink` would be spawned with `nsexec --keep-fd`, then communicate with target socket

The difference is the boundary of the "tunnel/proxy" component, IMO, using `fd://3` means the target process must be our subprocess, it's not a common suppose. And application listens on a network address or unix domain socket is much more common.

Reply:
If you think the `tunnel` is helpful (for debug, or other purpose), you can implement it with the help of nsexec and based on the same strategy described in this RPC. But I do think even if there is already a `tunnel` that we can use, inheriting the file descriptors is much better, in our use case.